web/src/components/pages/contribution/speak/audio-web.ts

```diff
@@ -183,15 +188,30 @@ export default class AudioWeb {
 
     return new Promise<void>((res: Function, rej: Function) => {
       this.chunks = [];
-      this.recorder.ondataavailable = (e: BlobEvent) => {
-        this.chunks.push(e.data);
-      };
-
-      this.recorder.onstart = (e: Event) => {
+      // Remove the old listeners.
+      this.recorder.removeEventListener('start', this.recorderListeners.start);
```

Comment:
Just out of curiosity, can you explain why you needed all the events to be on a new object instead of `this.recorder`? I skimmed the polyfill documentation but nothing jumped out at me and I'm sure there's a good reason for it.

Reply:
**TL;DR,** because:

1. `this.recorder` isn’t an object signature that we created or control.
2. There’s no standard DOM method to access a node’s event listeners.
3. The polyfill expects us to use `add/removeEventListener()` rather than setting `onStart` directly.

I think it’s cleaner to create a state object rather than add nonstandard properties (eg. `_listeners`) to a `MediaRecorder` instance.

---

Good question! This isn't exactly polyfill-specific, so some example code might help clarify:

```.js
// Goal: Add a click listener, then remove it a bit later.
const el = document.createElement('div');

////////////////////////////////
// Easy mode: shared closure. //
////////////////////////////////
addAndRemoveHandler('clicked');

function addAndRemoveHandler (msg) {
	const handler = () => alert(msg);
	el.addEventListener('click', handler);

	// Remove the event listener after 5 seconds. We can do this because we
	// still have access to the initial handler within this scope.
	window.setTimeout(() => {
		el.removeEventListener('click', handler);
	}, 5000);
}

///////////////////////////////////
// Hard mode: separate closures. //
///////////////////////////////////
addHandler('clicked');
window.setTimeout(removeHandler, 5000);

function addHandler (msg) {
	const handler = () => alert(msg);
	el.addEventListener('click', handler);
}

function removeHandler () {
	el.removeEventListener('click', /* uh oh… we can’t access
	                                   the handler from here. */);
}
```

There are many ways around this problem. Some common solutions are:

1. Easiest, but a bit of a hack: avoid `add/removeEventListener()`. Instead, set the `onClick` (or `onStart`, etc) property directly. This overwrites any listeners that existed beforehand.
2. Still pretty easy, not _e l e g a n t:_ create a state object in the parent scope to keep track of active handlers. That way, you can still reference the handler when it’s time to remove it.
3. Fancy: pass everything around as arguments and returns, don't track state at the parent level.

We’ve gone with option 1 until now by setting `onStart` directly. Buuuut, the polyfill is just using a `documentFragment` internally; it knows how to handle `add/removeEventListener()`, but it doesn’t know what to do with `onStart`.

Which leaves us with option 2 or 3. Option 3 is nice in theory, but would have been a ridiculous tangle with our current implementation.

So, option 2 it is. And I still haven’t even answered your question yet.

We could add a custom property to `this.recorder`, but it’s similar to adding a custom `_listeners` property to `el` in the example above. If browser vendors or the polyfill author later decide to add a `_listeners` property to the `MediaRecorder` instance, it will cause a bug. So, keeping track of state in our own separate object is safer (and hopefully somewhat clearer).

Reply:
Awesome, thanks for the clear explanation!

Reply:
Awesome, thanks for the clear explanation!